### ISSUE TITLE
ResourceApplyGradientDescent happens on CPU?

### ISSUE THREAD

**Issue by TobiasNorlund** [tensorflow/tensorflow#34047]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 19.10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.0.0
- Python version: 3.6.8
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA Version: 10.1
```bash
tf-docker /opt/project/src > nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2018 NVIDIA Corporation
Built on Sat_Aug_25_21:08:01_CDT_2018
Cuda compilation tools, release 10.0, V10.0.130
```
- GPU model and memory: GeForce GTX 1050, 4042MiB

I use the official tensorflow docker image: `tensorflow/tensorflow:2.0.0-gpu-py3`

**Describe the current behavior**
I experience fluctuating GPU utilization when I train a simple language model on artificial data using a custom training loop.
![Peek 2019-11-06 17-01](https://user-images.githubusercontent.com/2678217/68315470-197fa680-00b8-11ea-936d-b2c588259558.gif)

**Describe the expected behavior**
I would expect to get GPU utilization more stable at 100% (or very close to it), since the model is very simple and I expect all (heavy) ops to be able to run on the GPU.

**Code to reproduce the issue**
```python
import tensorflow as tf
#tf.debugging.set_log_device_placement(True)

VOCAB_SIZE = 32000
BATCH_SIZE = 20
LEARNING_RATE = 0.001

class SimpleModel(tf.keras.Model):
    def __init__(self, target_vocab_size):
        super(SimpleModel, self).__init__()
        self.d_model = 1024
        self.embedding = tf.keras.layers.Embedding(target_vocab_size, self.d_model)
        self.body = tf.keras.layers.Dense(self.d_model)

    def call(self, input, training):
        # input shape: (batch, seq_len)
        x = self.embedding(input)  # (batch, seq_len, d_model)
        x = self.body(x)
        logits = tf.matmul(x, self.embedding.embeddings, transpose_b=True)
        return logits


def get_dataset():
    ds = tf.data.Dataset.from_tensor_slices(tf.random.uniform((1000, 20), 0, VOCAB_SIZE, dtype=tf.int64))
    ds = ds.padded_batch(BATCH_SIZE, padded_shapes=[None])
    ds = ds.repeat()
    ds = ds.prefetch(2)
    return ds


def train(ds, model, optimizer):

    loss_object = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

    def calculate_loss(real, pred):
        loss_ = loss_object(real, pred)
        return tf.reduce_mean(loss_)

    @tf.function
    def train_step(batch):
        tar_inp = batch[:, :-1]
        tar_real = batch[:, 1:]

        with tf.GradientTape() as tape:
            logits = model(tar_inp, True)
            loss = calculate_loss(tar_real, logits)

        gradients = tape.gradient(loss, model.trainable_variables)
        optimizer.apply_gradients(zip(gradients, model.trainable_variables))
        return None

    #tf.summary.trace_on(graph=True, profiler=True)

    for i, batch in enumerate(ds):
        _ = train_step(batch)

        #if i == 10:
        #   tf.summary.trace_export(name="model_trace", step=0, profiler_outdir="/tmp/profiling/5")
        #   break


if __name__ == "__main__":
    ds = get_dataset()
    model = SimpleModel(VOCAB_SIZE)
    optimizer = tf.keras.optimizers.SGD(LEARNING_RATE)

    train(ds, model, optimizer)
```

**Other info / logs**
When inspecting the trace in Tensorboard I see two memcpys happening between training steps (green and blue boxes):
![Screenshot from 2019-11-06 17-16-15](https://user-images.githubusercontent.com/2678217/68316249-7039b000-00b9-11ea-9b44-88f3f58fa5e1.png)

Zooming in between two steps, I also see a ResourceApplyGradientDescent op happening on the CPU:
![Screenshot from 2019-11-06 17-19-48](https://user-images.githubusercontent.com/2678217/68316446-c73f8500-00b9-11ea-863a-b050eddeab8d.png)

Are ResourceApply* ops supposed to happen on the host (CPU)?

**Comment by gadagashwini** [comment]
@TobiasNorlund, Please refer this [link](https://www.tensorflow.org/guide/gpu). The document describes the GPU utilization. Thanks!

**Comment by TobiasNorlund** [comment]
Even if I explicitly pin the GPU to the op applying gradients, there is still no change. It is still performed on the CPU according to trace.
```
        with tf.device('/job:localhost/replica:0/task:0/device:GPU:0'):
            gradients = tape.gradient(loss, model.trainable_variables)
            optimizer.apply_gradients(zip(gradients, model.trainable_variables))
```

Is this to be expected?

**Comment by sanjoy** [comment]
To me the smoking gun here seems to be this line when I enable logging:

```
2019-12-27 10:09:44.089327: I tensorflow/core/common_runtime/eager/execute.cc:533] Executing op VarHandleOp in device /job:localhost/replica:0/task:0/device:CPU:0
```

and then in the graph we have:

```
node {
  name: "simple_model_embedding_embedding_lookup_readvariableop_resource"
  op: "_Arg"
  device: "/job:localhost/replica:0/task:0/device:CPU:0"
  attr {
    key: "T"
    value {
      type: DT_RESOURCE
    }
  }
  attr {
    key: "_handle_dtypes"
    value {
      list {
        type: DT_FLOAT
      }
    }
  }
  attr {
    key: "_handle_shapes"
    value {
      list {
        shape {
          dim {
            size: 32000
          }
          dim {
            size: 1024
          }
        }
      }
    }
  }
  attr {
    key: "index"
    value {
      i: 1
    }
  }
}

...

node {
  name: "SGD/SGD/update/ResourceApplyGradientDescent"
  op: "ResourceApplyGradientDescent"
  input: "simple_model_embedding_embedding_lookup_readvariableop_resource"
  input: "SGD/Identity_2"
  input: "AddN"
  device: "/job:localhost/replica:0/task:0/device:CPU:0"
  attr {
    key: "T"
    value {
      type: DT_FLOAT
    }
  }
  attr {
    key: "use_locking"
    value {
      b: true
    }
  }
}
```

It looks like in eager mode we create the resource variable on the CPU that causes the `ResourceApplyGradientDescent` to also run on the CPU.  I'm not sure why we're creating the resource variable on the CPU in the first place, the other variables are put on the GPU as expected.

Assigning to @qqfish

**Comment by sachinprasadhs** [comment]
Was able to run your code in Tensorflow GPU 2.5, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/5cada4be64e2596e22261cf99ca5343d/34047.ipynb) and let us know if you are still facing the issue. Thanks!

**Comment by aingo03304** [comment]
I am also struggling with this issue when I apply mixed-precision training. @TobiasNorlund Did you find out what the problem is?
